src/Compilers/Core/MSBuildTask/MvidReader.cs

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.BuildTasks
+{
+    public sealed class MvidReader : BinaryReader
+    {
+        public MvidReader(Stream stream) : base(stream)
+        {
+        }
+
+        public static Guid ReadAssemblyMvidOrEmpty(Stream stream)
+        {
+            try
+            {
+                var mvidReader = new MvidReader(stream);
+                return mvidReader.TryFindMvid();
+            }
+            catch (EndOfStreamException)
+            {
+            }
+
+            return Guid.Empty;
+        }
+
+        public Guid TryFindMvid()
+        {
+            Guid empty = Guid.Empty;
+
+            // DOS Header (64), DOS Stub (64), PE Signature (4), COFF Header (20), PE Header (224)
+            if (BaseStream.Length < 64 + 64 + 4 + 20 + 224)
+            {
+                return empty;
+            }
+
+            // DOS Header: PE (2)
+            if (ReadUInt16() != 0x5a4d)
+            {
+                return empty;
+            }
+
+            // DOS Header: Start (58)
+            Skip(58);
+
+            // DOS Header: Address of PE Signature
+            MoveTo(ReadUInt32());
```

Comment:
>MoveTo(ReadUInt32()); [](start = 12, length = 21)

It looks like we are moving to an unknown position without verifying if the destination makes sense.